include/net/downloader.h

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/**@file downloader.h
+ *
+ * @defgroup downloader Downloader
+ * @{
+ * @brief Client for downloading an object.
+ *
+ * @details The downloader provides APIs for:
+ *  - downloading an object from the server,
+ *  - receiving asynchronous event notifications on the download status.
+ */
+
+#ifndef __DOWNLOADER_H__
+#define __DOWNLOADER_H__
+
+#include <zephyr/kernel.h>
+#include <zephyr/types.h>
+#include <zephyr/net/coap.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define K_THREAD_STACK_MEMBER K_KERNEL_STACK_MEMBER
+
+/**
+ * @brief Downloader event IDs.
+ */
+enum downloader_evt_id {
+	/**
+	 * Event contains a fragment of data received from the server.
+	 * When using range requests the amount of data per fragment may be less than the
+	 * range requested.
+	 * The application may return any non-zero value to stop the download.
+	 */
+	DOWNLOADER_EVT_FRAGMENT,
+	/**
+	 * An error has occurred during download and
+	 * the connection to the server has been lost.
+	 *
+	 * Error reason may be one of the following:
+	 * - ECONNRESET: Socket error, peer closed connection
```

Comment:
The error reason in the event is negative, it would be better if this would visible here.